src/Agent.Worker/TaskDefinitionManager.cs

```diff
@@ -0,0 +1,291 @@
+                    {
+                        //81920 is the default used by System.IO.Stream.CopyTo and is under the large object heap threshold (85k). 
+                        await result.CopyToAsync(fs, 81920, executionContext.CancellationToken);
+                        await fs.FlushAsync(executionContext.CancellationToken);
+                    }
+                }
+
+                System.IO.Compression.ZipFile.ExtractToDirectory(zipFile, tempDirectory);
+                File.Delete(zipFile);
+                Directory.CreateDirectory(Path.GetDirectoryName(destDirectory));
+                Directory.Move(tempDirectory, destDirectory);
+                Trace.Info("Finished getting task.");
+            }
+            finally
+            {
+                try
+                {
+                    //sometimes the temp folder is not deleted -> wipe it
+                    if (Directory.Exists(tempDirectory))
+                    {
+                        Trace.Verbose("Deleting task temp folder: {0}", tempDirectory);
+                        Directory.Delete(tempDirectory, true);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    //it is not critical if we fail to delete the temp folder
+                    Trace.Warning("Failed to delete temp folder '{0}'. Exception: {1}", tempDirectory, ex);
+                    executionContext.Warning(StringUtil.Loc("FailedDeletingTempDirectory0Message1", tempDirectory, ex.Message));
+                }
+            }
+        }
+
+        private string GetDirectory(TaskReference task)
+        {
+            ArgUtil.NotEmpty(task.Id, nameof(task.Id));
+            ArgUtil.NotNull(task.Name, nameof(task.Name));
+            ArgUtil.NotNullOrEmpty(task.Version, nameof(task.Version));
+            return Path.Combine(
+                IOUtil.GetTasksPath(HostContext),
+                $"{task.Name}_{task.Id}",
+                task.Version);
+        }
+
+        public async Task DownloadAsync(IExecutionContext executionContext, IEnumerable<TaskInstance> tasks)
+        {
+            ArgUtil.NotNull(executionContext, nameof(executionContext));
+            ArgUtil.NotNull(tasks, nameof(tasks));
+
+            //remove duplicate and disabled tasks
+            IEnumerable<TaskInstance> uniqueTasks =
+                from task in tasks
+                where task.Enabled
+                group task by new
+                    {
+                        task.Id,
+                        task.Version
+                    }
+                into taskGrouping
+                select taskGrouping.First();
+            foreach (TaskInstance task in uniqueTasks)
+            {
+                await DownloadAsync(executionContext, task);
+            }
+        }
+
+        public Definition Load(TaskReference task)
```

Comment:
This is the new method to load the definition data. TaskRunner calls this so it can choose (and initialize) the appropriate handler.